File: python/array/4Sum.py
#mistakes were made:
#0. should use dict to store index rather than value
#in this case, we won't use a same number twice
class Solution:
    # @return a list of lists of length 4, [[val1,val2,val3,val4]]
    def fourSum(self, num, target):
        d = {}
        num.sort()
        result = set()
        result_list = []
        for i in xrange(len(num)):
            for j in xrange(i+1, len(num)):
                curr = num[i] + num[j]
                if curr not in d:
                    d[curr] = [(i, j)]
                else:
                    d[curr].append((i, j))

        for i in xrange(len(num)):
            for j in xrange(i+1, len(num)):
                temp = target - num[i] - num[j]
                if temp in d:
                    for k in xrange(len(d[temp])):
                        if d[temp][k][0] > j:
                            result.add((num[i], num[j], num[d[temp][k][0]], num[d[temp][k][1]]))
        for i in result:
            result_list.append(list(i))

        return result_list

if __name__ == "__main__":
    solution = Solution()

    print solution.fourSum([-2, -1, 0, 0, 1, 2], 0)

File: c++/array/4Sum.cpp
class Solution {
public:
    std::vector<std::vector<int> > fourSum(std::vector<int> &num, int target) {
        std::unordered_map<int, std::vector<std::pair<int, int>>> dictionary;
        std::sort(num.begin(), num.end());
        std::set<std::vector<int>> result;
        
        for (int i = 0; i < int(num.size()); i ++) {
            for (int j = i; j < int(num.size()); j ++) {
                dictionary[num[i]+num[j]].push_back(std::make_pair(i, j));
            }
        }
        
        for (int i = 0; i < int(num.size()); i ++) {
            if (i != 0 && num[i] == num[i-1]) {
                continue;
            }
            for (int j = i + 1; j < int(num.size()); j ++) {
                int curr = num[i] + num[j];
                if (dictionary.find(target-curr) != dictionary.end()) {
                    std::vector<std::pair<int, int>> sum2 = dictionary[target-curr];
                    for (int k = 0; k < sum2.size(); k ++) {
                        if (sum2[k].first > j) {
                            std::vector<int> item;
                            item.push_back(num[i]);
                            item.push_back(num[j]);
                            item.push_back(num[sum2[k].first]);
                            item.push_back(num[sum2[k].second]);
                            result.insert(item);
                        }
                    }
                }
            }
        }
        
        std::vector<std::vector<int>> resultList;
        
        std::set<std::vector<int>>::iterator set_iter;
        for (set_iter = result.begin(); set_iter != result.end(); set_iter++) {
            resultList.push_back(*set_iter);
        }
        
        return resultList;
    }
};
